### PR TITLE
lammps-mpi: Use a generic extraBuildInputs instead of withMpi

### PR DESCRIPTION
**Pull request by doronbehar**
Add support for `extraCmakeFlags` as well and add `pkg-config` to `nativeBuildInputs`.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by doronbehar** [comment]
I removed the `withMpi` and `mpi` expression arguments.

File: pkgs/applications/science/molecular-dynamics/lammps/default.nix
Comment (doronbehar) [reply pkgs/applications/science/molecular-dynamics/lammps/default.nix:59]: Hmm.. well indeed this is a breaking change. However, you can still access the `mpi` that was used when compiling `lammps` via `lammps.passthru.extraBuildInputs`. It's a list and not an attribute, but still.. Are you worried about the incompatibility in general? Perhaps we could `throw` an error message for those that relied on that attribute.

File: pkgs/applications/science/molecular-dynamics/lammps/default.nix
Comment (doronbehar) [reply pkgs/applications/science/molecular-dynamics/lammps/default.nix:42]: > Is there a particular motivation why you want to add MPI in that fashion, and deviate from the widely used 'withMpi' pattern?

I just don't see a reason why we would treat `mpi` differently then other "packages" and optional dependencies for this package. I want all of the switches and knobs of this package to have the same interface.

> It requires, for example, additional binaries from the mpi package, such as mpirun.

I understand. If people use `lammps` in their development environment, now they can simply add `lammps-mpi.passthru.extraBuildInputs` to their devShell's `buildInputs`, and get the same `mpi` as that of `lammps-mpi`.

Comment (markuskowa) [reply pkgs/applications/science/molecular-dynamics/lammps/default.nix:42]: > I just don't see a reason why we would treat mpi differently then other "packages" and optional dependencies for this package. I want all of the switches and knobs of this package to have the same interface.

It's probably a matter of taste how special mpi is. I would argue that MPI changes the runtime environment, which requires extra tools to even run the binary and thus has a special standing amongst the inputs. 

> I understand. If people use lammps in their development environment, now they can simply add lammps-mpi.passthru.extraBuildInputs to their devShell's buildInputs, and get the same mpi as that of lammps-mpi.

Yes, that would work. When mpi is in `propagatedBuildInputs` it would even work without special intervention.

Using `extraBuildInputs` is a bit of a paradigm shift, that is not widely used (but an interesting approach). Wouldn't it be better to have uniform solution?

Comment (doronbehar) [reply pkgs/applications/science/molecular-dynamics/lammps/default.nix:42]: > Using `extraBuildInputs` is a bit of a paradigm shift, that is not widely used (but an interesting approach). Wouldn't it be better to have uniform solution?

What kind of a solution would you like to see? I mean, unifying which ideas?

> Yes, that would work. When mpi is in `propagatedBuildInputs` it would even work without special intervention.

We could also add `propagatedBuildInputs` as an argument to the expression... 

With a good message in `passthru.mpi = throw "...";` and a release notes entry, perhaps this is good enough for `nixos-unstable`?

TBH I feel very comfortable introducing breaking changes to the interface of this package, as it was very outdated no long ago, and I doubt there were many if any, users relying on the `passthru.mpi` interface.

**Review by markuskowa — APPROVED** [review]
If you like this solution, then I am okay with it.

We (@sheepforce and me) have been thinking about how to deal with MPI packages for a while. The pattern that appeared most practical to us, is `withMpi`, `propagatedBuildInputs`, and `passthru.mpi`, and we have started using it in a bunch of packages. However, that does not mean, that it is the only answer. Since we do not have a well unified way in nix/nixpkgs to define and discover package options, there is room for creativity here.

**Comment by doronbehar** [comment]
Thanks for your approval @markuskowa . I added a `passthru.mpi.throw` message. If you wish, I can wait for you a bit for your approval on it's phrasing.

**Pull request merged**
